execution.bs

```diff
@@ -3173,67 +3196,91 @@ Insert this section as a new subclause between Header `<stop_token>` synopsis <b
+    token type that does not allow stopping.
 
 <pre>
-template&lt;typename T>
+template&lt;class T>
   concept stoppable_token =
-    copy_constructible&lt;T> &&
-    move_constructible&lt;T> &&
-    is_nothrow_copy_constructible_v&lt;T> &&
-    is_nothrow_move_constructible_v&lt;T> &&
+    copyable&lt;T> &&
     equality_comparable&lt;T> &&
-    requires (const T& token) {
-      { token.stop_requested() } noexcept -> boolean-testable;
-      { token.stop_possible() } noexcept -> boolean-testable;
+    requires (const T t) {
+      { T{t} } noexcept;
+      { t.stop_requested() } noexcept -> <i>boolean-testable</i>;
+      { t.stop_possible() } noexcept -> <i>boolean-testable</i>;
       typename <i>check-type-alias-exists</i>&lt;T::template callback_type>;
     };
 
-template&lt;typename T, typename CB, typename Initializer = CB>
+template&lt;class T, class CB, class Initializer = CB>
   concept stoppable_token_for =
     stoppable_token&lt;T> &&
     invocable&lt;CB> &&
-    requires {
-      typename T::template callback_type&lt;CB>;
-    } &&
     constructible_from&lt;CB, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, T, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, T&, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, const T, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, const T&, Initializer>;
+    requires (const T t) {
+      typename stop_callback_for_t&lt;T, CB>;
+      []&lt;class T>(T&& t)
+        requires constructible_from&lt;stop_callback_for_t&lt;T, CB>, U, Initializer>
+      {}(t);
+    };
 
-template&lt;typename T>
+template&lt;class T>
   concept unstoppable_token =
     stoppable_token&lt;T> &&
     requires {
-      { T::stop_possible() } -> boolean-testable;
-    } &&
-    (!T::stop_possible());
+      { bool_constant&lt;T::stop_possible()>{} } -> same_as&lt;false_type>;
+    };
 </pre>
 
-2. Let `t` and `u` be distinct object of type `T`. The type `T` models `stoppable_token` only if:
+<div class="ed-note">LWG directed me to replace `T::stop_possible()` with `t.stop_possible()` because
+of the recent `constexpr` changes in [[P2280r2|P2280R2]]. However, even with those changes, a nested
+requirement like `requires (!t.stop_possible())`, where `t` is an argument in the requirement-parameter-list, is ill-formed according to
+<a href="http://eel.is/c++draft/expr.prim.req#nested-2.sentence-1">[expr.prim.req.nested/p2]</a>:
+
+> A local parameter shall only appear as an unevaluated operand within the constraint-expression.
+
+This is the subject of core issue [[cwg2517|2517]].
+</div>
+
+2. Let `t` and `u` be distinct, valid objects of type `T`. The type `T` models `stoppable_token` only if:
 
-    1. All copies of a `stoppable_token` reference the same logical shared stop state and shall report values consistent with each other.
```

Comment:
Do we still say anywhere that it is a requirement that a copy of a stoppable-token must refer to the same shared stop-state that the original referred to?

Reply:
copyable in generał require copies to be equal.